crates/apollo-compiler/src/coordinate.rs

```diff
@@ -0,0 +1,684 @@
+//! Parsing and printing for schema coordinates as described in [the RFC].
+//!
+//! Schema coordinates uniquely point to an item defined in a schema.
+//!
+//! [the RFC]: https://github.com/graphql/graphql-wg/blob/main/rfcs/SchemaCoordinates.md
+
+use crate::ast::InvalidNameError;
+use crate::ast::Name;
+use crate::schema::Component;
+use crate::schema::DirectiveDefinition;
+use crate::schema::EnumValueDefinition;
+use crate::schema::ExtendedType;
+use crate::schema::FieldDefinition;
+use crate::schema::InputValueDefinition;
+use crate::schema::NamedType;
+use crate::schema::Schema;
+use crate::Node;
+use std::fmt;
+use std::str::FromStr;
+
+/// Create a static schema coordinate at compile time.
+///
+/// ```rust
+/// use apollo_compiler::coord;
+///
+/// assert_eq!(coord!(@directive).to_string(), "@directive");
+/// assert_eq!(coord!(@directive(arg:)).to_string(), "@directive(arg:)");
+/// assert_eq!(coord!(Type).to_string(), "Type");
+/// assert_eq!(coord!(Type.field).to_string(), "Type.field");
+/// assert_eq!(coord!(Type.field(arg:)).to_string(), "Type.field(arg:)");
+/// assert_eq!(coord!(EnumType.ENUM_VALUE).to_string(), "EnumType.ENUM_VALUE");
+/// ```
+#[macro_export]
+macro_rules! coord {
+    ( @ $name:ident ) => {
+        $crate::coordinate::DirectiveCoordinate {
+            directive: $crate::name!($name),
+        }
+    };
+    ( @ $name:ident ( $arg:ident : ) ) => {
+        $crate::coordinate::DirectiveArgumentCoordinate {
+            directive: $crate::name!($name),
+            argument: $crate::name!($arg),
+        }
+    };
+    ( $name:ident ) => {
+        $crate::coordinate::TypeCoordinate {
+            ty: $crate::name!($name),
+        }
+    };
+    ( $name:ident . $attribute:ident ) => {
+        $crate::coordinate::TypeAttributeCoordinate {
+            ty: $crate::name!($name),
+            attribute: $crate::name!($attribute),
+        }
+    };
+    ( $name:ident . $field:ident ( $arg:ident : ) ) => {
+        $crate::coordinate::FieldArgumentCoordinate {
+            ty: $crate::name!($name),
+            field: $crate::name!($field),
+            argument: $crate::name!($arg),
+        }
+    };
+}
+
+mod sealed {
+    pub trait Sealed {}
+}
+
+/// Provides [`schema.lookup(&coord)`][Schema::lookup] for any schema coordinate.
+pub trait SchemaLookup: sealed::Sealed {
```

Comment:
This trait only exists to support the `Schema::lookup` method, right? I feel it and the method don’t bring a lot of value, compared to separate `lookup` inherent methods on the various coordinate types. Anything that could be generic can use the enum instead.